### PR TITLE
Wrap the logger used by replay-driver to markup log events

### LINKED ISSUE #3902: Add a way in telemetry to tell if replay driver is being used.
When looking at the current telemetry, in some case during container load, we see that the OdspDriver try to fetch a whole bunch of snapshot via GetVersions before going down the "ObtainSnapshot" route.  It is likely that GetVersions is called with count != 1.

The only case that I know of that would happen is the FluidDebugger and replay driver, which we should discounted from our real life scenario perf measurement.

We can create a wrapped logger in ReplayDocumentServiceFactory.createDocumentService with 'isReplay: true" property to tag and identify all event from the driver that it is wrapped with a replay driver.

### PR DESCRIPTION
**Pull request by markfields**
Fixes #3902 

This allows us to mark up all events with isReplay: true so we can separate these logs from other logs.

### CODE REVIEW

**Review by christiango — APPROVED** [review]

File: packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts [hunk 1]
```diff
@@ -5,11 +5,25 @@
...
+ * Wrapper logger that adds isReplay: true to each event.
+ * ReplayDriver is used in testing/debugging scenarios, so we want to be able to filter these events out sometimes
+ */
+class ReplayDriverTelemetryLogger implements ITelemetryBaseLogger {
+    constructor(
+        private readonly logger: ITelemetryBaseLogger,
+    ) { }
+
+    send(event: ITelemetryBaseEvent): void {
+        this.logger.send({ ...event, isReplay: true });
```
Comment (markfields) [inline packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts:23]: Ohhh... Just looking at ChildLogger code, should probably use that :P

File: packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts [hunk 2]
```diff
@@ -5,11 +5,25 @@
...
+import { ITelemetryBaseEvent, ITelemetryBaseLogger } from "@fluidframework/common-definitions";
 import { ReplayController } from "./replayController";
 import { ReplayControllerStatic } from "./replayDocumentDeltaConnection";
 import { ReplayDocumentService } from "./replayDocumentService";
 
+/**
+ * Wrapper logger that adds isReplay: true to each event.
+ * ReplayDriver is used in testing/debugging scenarios, so we want to be able to filter these events out sometimes
+ */
+class ReplayDriverTelemetryLogger implements ITelemetryBaseLogger {
```
Comment (vladsud) [inline packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts:17]: I do not see a reason for this class to exists - can you please use this instead?
```
ChildLogger.create(logger, undefined, { isReplay: true }) 
```

**Comment by markfields** [comment]
@Curtisman @vladsud - I'm also looking at `FluidDebugger.createFromService` - I don't see an obvious way to inject a logger in that flow, is that right?

And coming at it from the other angle - if I search for "logger" under replay-driver, I only find one other hit, which is n/a (`StaticStorageDocumentServiceFactory.createContainer` which just throws "Not implemented"). So it seems like even after this change there's going to be scenarios using replay driver that don't get the logging marked up.

File: packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts (hunk 2 above)
Comment (markfields) [reply packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts:17]: Yep just noticed that.  Not sure why my comment didn't appear for you...

File: packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts [hunk 3]
```diff
@@ -39,8 +53,9 @@ export class ReplayDocumentServiceFactory implements IDocumentServiceFactory {
         resolvedUrl: IResolvedUrl,
         logger?: ITelemetryBaseLogger,
     ): Promise<IDocumentService> {
+        const replayLogger = logger && new ReplayDriverTelemetryLogger(logger);
```
Comment (vladsud) [inline packages/drivers/replay-driver/src/replayDocumentServiceFactory.ts:56]: Note that you do not need to test for undefined when using ChildLogger.create() - it's happy with creating null logger on demand

**Comment by vladsud** [comment]
There is no way to inject your logger with FluidDebugger.createFromService flow, as the underlying service is already initialized and already uses some logger. This path is not used, I believe, but it's super valuable for local testing (I just wrap service in container.ts with it when making any changes to test it with any existing workflow).

I think with latest way how we create driver (now it's created by container, it used to be created by loader, I think), we can remove this API - we always operate with factory.

**Review by vladsud — APPROVED** [review]

**Pull request merged**